### PR TITLE
WI senate vote adding function re-written

### LINKED ISSUE #1909: WI: All 2017 individual roll call votes are missing for Senate, some Assembly too
The Senate currently has no individual roll call votes recorded in 2017.

### PR DESCRIPTION
**Pull request by In-vincible**
resolves #1909
Will check for assembly votes too.

### CODE REVIEW

**Comment by In-vincible** [comment]
close #1909

**Comment by mileswwatkins** [comment]
It's helpful if your PR title starts with the state abbreviation, for reference; eg `WI: …`

**Comment by mileswwatkins** [comment]
Also, is this ready for review?

**Comment by In-vincible** [comment]
Yes

File: openstates/wi/bills.py [hunk 1]
```diff
@@ -375,12 +368,24 @@ def add_house_votes(self, vote, url):
...
             name = td.text_content()
             for vote_td in td.xpath('./preceding-sibling::td'):
                 if vote_td.text_content() == 'Y':
                     vote.vote('yes', name)
+                    yes_names_count += 1
                 elif vote_td.text_content() == 'N':
                     vote.vote('no', name)
+                    no_names_count += 1
                 elif vote_td.text_content() == 'NV':
                     vote.vote('other', name)
```
Comment (mileswwatkins) [inline openstates/wi/bills.py]: It looks like this category is actually `not voting`, not `other`, right?

File: openstates/wi/bills.py [hunk 2]
```diff
@@ -329,36 +329,29 @@ def add_senate_votes(self, vote, url):
...
+            name_counts[vote_type] = len(names)
+
+            for name in names:
+                vote.vote(vote_type, name.strip())
+
+        for vote_type in vote_types:
+            vote.set_count(vote_type, vote_counts[vote_type])
+
+        if name_counts != vote_counts:
+            self.error("Yes votes and number of Names doesn't match")
```
Comment (mileswwatkins) [inline openstates/wi/bills.py]: Wrong error message?

File: openstates/wi/bills.py [hunk 3]
```diff
@@ -329,36 +329,29 @@ def add_senate_votes(self, vote, url):
...
-                        vote.vote('other', name.strip())
-                        other_count += 1
-            elif text.strip():
-                raise ValueError('unexpected block in vote')
-
-        if yes_count or no_count or other_count:
-            vote.set_count('yes', yes_count)
-            vote.set_count('no', no_count)
-            vote.set_count('other', other_count)
+        for index in range(0, 5, 2):
```
Comment (mileswwatkins) [inline openstates/wi/bills.py]: This seems like weird construction, both non-standard Python and hard to read (for future maintainers). Could you do something like this instead?

- XPath for `<td>` that has text of `(AYES|NAYS|NOT VOTING) - \d+`
- get the names from the `<table>` in the [following-sibling](https://www.w3schools.com/xml/xpath_axes.asp) `<td>`

File: openstates/wi/bills.py [hunk 4]
```diff
@@ -329,36 +329,29 @@ def add_senate_votes(self, vote, url):
             return
 
         doc = lxml.html.fromstring(html)
+        trs = doc.xpath('//table[@class="senate"]/tbody/tr')
 
-        yes_count = no_count = other_count = 0
+        vote_types = ['yes', 'no', 'other']
```
Comment (mileswwatkins) [inline openstates/wi/bills.py]: The third type is `not voting`, right?

File: openstates/wi/bills.py (hunk 2 above)
Comment (In-vincible) [reply openstates/wi/bills.py]: +1

File: openstates/wi/bills.py (hunk 3 above)
Comment (In-vincible) [reply openstates/wi/bills.py]: 100% agreed, doing it.

File: openstates/wi/bills.py (hunk 4 above)
Comment (In-vincible) [reply openstates/wi/bills.py]: I am not really sure, absent, excused were not given, are they are taking all others in ```not voting``` only, that's why I thought maybe other would be right to use here. But yeah ```not voting``` is mentioned here. I should with ```not voting``` then?

Comment (mileswwatkins) [reply openstates/wi/bills.py]: Yes, if the state labels them as "Not Voting" then we should use that Open Civic Data category.

**Comment by In-vincible** [comment]
@mileswwatkins ready for next round of review.

File: openstates/wi/bills.py [hunk 5]
```diff
@@ -350,7 +353,7 @@ def add_senate_votes(self, vote, url):
             vote.set_count(vote_type, vote_counts[vote_type])
 
         if name_counts != vote_counts:
-            self.error("Yes votes and number of Names doesn't match")
+            self.error("Names not extracted correctly")
             raise ValueError("Vote Count and number of Names don't match")
```
Comment (mileswwatkins) [inline openstates/wi/bills.py]: Are `self.error` and `raise …` redundant?

Comment (mileswwatkins) [reply openstates/wi/bills.py]: Looks like there are a couple cases further down in the code, too.

Comment (In-vincible) [reply openstates/wi/bills.py]: I will delete self.error then!

Comment (In-vincible) [reply openstates/wi/bills.py]: I personally think, we should throw error, as we will get notified for that, and hence we can act on it.

**Comment by In-vincible** [comment]
@mileswwatkins I have removed ```self.error``` line, thought should stick with raise error, so that we get notified in case of descrepancies.

**Pull request merged**